nomad/state/state_store_variables.go

```diff
@@ -511,3 +518,83 @@ func (s *StateStore) VariablesQuotaByNamespace(ws memdb.WatchSet, namespace stri
 	quotaUsed := raw.(*structs.VariablesQuota)
 	return quotaUsed, nil
 }
+
+// VarLockAcquire is the method used to append a lock to a variable, if the
+// variable doesn't exists, it is created.
+func (s *StateStore) VarLockAcquire(idx uint64,
+	req *structs.VarApplyStateRequest) *structs.VarApplyStateResponse {
+	tx := s.db.WriteTxn(idx)
+	defer tx.Abort()
+
+	// Try to fetch the variable.
+	raw, err := tx.First(TableVariables, indexID, req.Var.Namespace, req.Var.Path)
+	if err != nil {
+		return req.ErrorResponse(idx, fmt.Errorf("variable lookup failed: %v", err))
+	}
+
+	// If the variable exist, we must make sure it doesn't hold a lock already
+	if raw != nil {
+		sv := raw.(*structs.VariableEncrypted)
+		if sv.VariableMetadata.Lock != nil {
+			return req.ErrorResponse(idx, errVarAlreadyLocked)
+		}
+
+		// Update the data to avoid overwriting the variable's items while creating the lock
+		req.Var.VariableData = sv.VariableData
```

Comment:
I'm not sure I understand this bit. I'm looking at the [NMD-178 (internal link)](https://docs.google.com/document/d/1OHUUNRMewMCFPi6z4VFUs0reE_AELE9zpJUg8I-rkuk/edit#heading=h.t91xls314i) RFC and it looks like the intent was that we'd use the existing Upsert API for the lock upserts too. Which means that this would drop any updates to the variable contents that were included with the lock upsert.

It'd definitely be weird for a user to do that, but it doesn't seem like the API prevents it and it would silently drop their changes.

That also makes me wonder how we'd handle someone locking _and_ setting the CAS flags.

Reply:
it was indeed the intent, reuse the variable upserts for both renew and release lock, because the operations end up being a write to the variable in the end, but it does leaves us with some interesting choices on the usability side.

We have not talked about locks + CAS so far,  but it is possible for the user to end up modifying or deleting the lock while upserting the variable, any ideas on that one?

Reply:
> it is possible for the user to end up modifying or deleting the lock while upserting the variable, any ideas on that one?

In which case, maybe we don't actually want to have a separate state store "operation" for this feature and instead add handling of the lock to the existing `VarSet` and `VarSetCAS` operations. 

We could also treat the lock ID of a locked variable like a CAS protection as well. That way:
* If you try to change a locked variable w/o having the lock ID, you'll get a conflict
* If you try to change the locked variable while having the lock ID but having the wrong CAS index, you'll get a conflict
* If you try to change the locked variable while having the lock ID and the right CAS index, it's updated
* If you try to change the locked variable while having the lock ID and no CAS index, it's updated

Reply:
The acquire or release a lock, some extra validations need to be done, inside one transaction. VarSet and VarSetCAS don't have them, forcing us to create a new operation.

Reply:
lock ID check added to CAS upsert and both simple and CAS delete :D